File: src/pymat/curves.py
"""Piecewise-linear property curves.

Two flavours, one interpolation rule:

- `TempCurve` — temperature-dependent properties (#148). Knots are
  `(temps_K, values)`.
- `WavelengthCurve` — wavelength-dependent optical properties (#243).
  Knots are `(wavelengths_nm, values)`.

Out-of-range abscissae are CLAMPED, not extrapolated — engineering data
extrapolated beyond its measured range is a lie, and clamping is
conservative and visibly wrong rather than subtly wrong (per ADR-0003 §2
edge-case table). ADR-0004 §4 extends the same rule to wavelength: a
dispersion table measured over 400-700 nm says nothing about 250 nm, and
a Sellmeier fit evaluated outside its stated validity range is worse than
useless because it looks like data.

Validation is at construction (and therefore at TOML load) — unsorted or
mismatched-length arrays raise `ValueError` immediately, not at query
time. Empty curves raise as well.

Used by sibling fields like `<prop>_curve: Optional[TempCurve]` and by
the wavelength accessors (`n_at`, `absorption_length_at`, `emission_at`)
on `OpticalProperties`.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any, List, Optional, Sequence, Tuple

logger = logging.getLogger(__name__)


def _validate_knots(xs: Sequence[float], ys: Sequence[float], cls: str, x_name: str) -> None:
    """Shared knot validation for every piecewise-linear curve.

    Raises `ValueError` on empty, length-mismatched, or non-strictly-ascending
    abscissae. Equal-adjacent knots make interpolation ambiguous; reject them.
    """
    if not xs:
        raise ValueError(f"{cls} requires at least one knot")
    if len(xs) != len(ys):
        raise ValueError(f"{cls} {x_name} and values must be same length: {len(xs)} vs {len(ys)}")
    for a, b in zip(xs, xs[1:]):
        if not a < b:
            raise ValueError(f"{cls} {x_name} must be strictly sorted ascending; got {list(xs)}")


def _interp_clamped(
    xs: Sequence[float], ys: Sequence[float], x: float, cls: str, x_name: str, unit: str
) -> float:
    """Evaluate a piecewise-linear curve at `x`, clamping outside the knot range."""
    if x <= xs[0]:
        if x < xs[0]:
            logger.debug(
                "%s: %s=%s %s below min knot %s %s; clamping", cls, x_name, x, unit, xs[0], unit
            )
        return ys[0]
    if x >= xs[-1]:
        if x > xs[-1]:
            logger.debug(
                "%s: %s=%s %s above max knot %s %s; clamping", cls, x_name, x, unit, xs[-1], unit
            )
        return ys[-1]
    for i in range(len(xs) - 1):
        x0, x1 = xs[i], xs[i + 1]
        if x0 <= x <= x1:
            y0, y1 = ys[i], ys[i + 1]
            return y0 + (x - x0) / (x1 - x0) * (y1 - y0)
    # Unreachable — the clamp branches above cover every case.
    raise RuntimeError(f"{cls}: failed to bracket {x_name}={x}")  # pragma: no cover


def _require_table(raw: Any, cls: str) -> dict:
    if not isinstance(raw, dict):
        raise ValueError(f"{cls} TOML must be a table, got {type(raw).__name__}")
    return raw


@dataclass(frozen=True)
class TempCurve:
    """Piecewise-linear temperature-dependent property curve.

    Attributes:
        temps_K: Strictly-sorted ascending temperatures in Kelvin.
        values: Property values at each knot, same length as `temps_K`.
    """

    temps_K: List[float]
    values: List[float]

    def __post_init__(self) -> None:
        _validate_knots(self.temps_K, self.values, "TempCurve", "temps_K")

    def interpolate(self, temp_K: float) -> float:
        """Evaluate the curve at `temp_K`. Out-of-range clamps to nearest knot."""
        return _interp_clamped(self.temps_K, self.values, temp_K, "TempCurve", "T", "K")

    @classmethod
    def from_toml(cls, raw: Any) -> "TempCurve":
        """Build from `{temps_K = [...], values = [...]}` TOML inline-table."""
        raw = _require_table(raw, "TempCurve")
        if "temps_K" not in raw or "values" not in raw:
            raise ValueError(f"TempCurve TOML missing 'temps_K' or 'values': {raw}")
        return cls(temps_K=list(raw["temps_K"]), values=list(raw["values"]))


# Value-column spellings accepted by `WavelengthCurve.from_toml` when no
# explicit `value_key` is given, in priority order. The non-canonical ones
# exist because the structured optical slots shipped in #153 / #164 before
# this primitive did, and their on-disk shape is load-bearing:
#   refractive_index_dispersion = {wavelengths_nm = [...], n = [...]}
#   emission_spectrum           = {wavelengths_nm = [...], intensities = [...]}
_WL_VALUE_KEYS: Tuple[str, ...] = ("values", "n", "intensities")


@dataclass(frozen=True)
class WavelengthCurve:
    """Piecewise-linear wavelength-dependent optical property curve (#243).

    The wavelength twin of `TempCurve`. Same clamping contract, same
    load-time validation. Wavelengths are always nanometres — the schema
    spells the abscissa `wavelengths_nm` everywhere, so there is no unit
    ambiguity to resolve at load.

    Attributes:
        wavelengths_nm: Strictly-sorted ascending wavelengths in nm.
        values: Property values at each knot, same length as `wavelengths_nm`.
    """

    wavelengths_nm: List[float]
    values: List[float]

    def __post_init__(self) -> None:
        _validate_knots(self.wavelengths_nm, self.values, "WavelengthCurve", "wavelengths_nm")

    def interpolate(self, wavelength_nm: float) -> float:
        """Evaluate at `wavelength_nm`. Out-of-range clamps to nearest knot."""
        return _interp_clamped(
            self.wavelengths_nm, self.values, wavelength_nm, "WavelengthCurve", "lambda", "nm"
        )

    @property
    def range_nm(self) -> Tuple[float, float]:
        """`(min, max)` of the measured range — the span outside which
        `interpolate` clamps. Downstream resamplers (e.g. strata's parquet
        emitter) use this to record where the data actually stops."""
        return (self.wavelengths_nm[0], self.wavelengths_nm[-1])

    @classmethod
    def from_toml(cls, raw: Any, value_key: Optional[str] = None) -> "WavelengthCurve":
        """Build from a `{wavelengths_nm = [...], <values> = [...]}` table.

        Args:
            raw: The TOML inline-table.
            value_key: Explicit name of the ordinate column. When omitted,
                the first of `values` / `n` / `intensities` present in the
                table wins. Two or more present at once is a hard error —
                silently picking one would make the file's meaning depend
                on this function's internals.
        """
        raw = _require_table(raw, "WavelengthCurve")
        if "wavelengths_nm" not in raw:
            raise ValueError(f"WavelengthCurve TOML missing 'wavelengths_nm': {raw}")

        if value_key is not None:
            if value_key not in raw:
                raise ValueError(f"WavelengthCurve TOML missing {value_key!r}: {raw}")
            key = value_key
        else:
            present = [k for k in _WL_VALUE_KEYS if k in raw]
            if not present:
                accepted = ", ".join(repr(k) for k in _WL_VALUE_KEYS)
                raise ValueError(
                    f"WavelengthCurve TOML has no value column (accepted: {accepted}): {raw}"
                )
            if len(present) > 1:
                raise ValueError(
                    f"WavelengthCurve TOML is ambiguous — multiple value columns "
                    f"{present}; pass value_key explicitly: {raw}"
                )
            key = present[0]

        return cls(wavelengths_nm=list(raw["wavelengths_nm"]), values=list(raw[key]))

File: mat-rs/src/curves.rs
//! Piecewise-linear property curves — the Rust mirror of `pymat.curves`.
//!
//! One type covers both axes. On the Python side `TempCurve` and
//! `WavelengthCurve` are distinct so that `_at(T)` and `n_at(lambda)` cannot be
//! confused at a call site; here the axis is fixed by the field the curve hangs
//! off, so a single `Curve` with a documented abscissa unit is enough.
//!
//! **Out-of-range abscissae are CLAMPED, not extrapolated.** This mirrors
//! ADR-0003 §2 / ADR-0004 §4: data extrapolated past its measured range is a
//! lie, and clamping is conservative and visibly wrong rather than subtly
//! wrong. Use [`Curve::range`] to find out where the measurement actually
//! stops — a resampler that silently clamps a 128-node grid onto a curve
//! measured over 60 nm has fabricated the other 68 nodes.

/// A piecewise-linear curve over a strictly-ascending abscissa.
///
/// `xs` is Kelvin for temperature curves and nanometres for wavelength curves.
#[derive(Debug, Clone, PartialEq)]
pub struct Curve {
    xs: Vec<f64>,
    ys: Vec<f64>,
}

impl Curve {
    /// Build a curve, validating the knots.
    ///
    /// Returns `None` when the knots are empty, length-mismatched, or not
    /// strictly ascending — the same three conditions that raise `ValueError`
    /// at load time on the Python side.
    pub fn new(xs: Vec<f64>, ys: Vec<f64>) -> Option<Self> {
        if xs.is_empty() || xs.len() != ys.len() {
            return None;
        }
        // `partial_cmp` rather than `>=` so NaN is rejected too: `a >= b` is
        // false for NaN, which would let a NaN knot through validation.
        if xs
            .windows(2)
            .any(|w| !matches!(w[0].partial_cmp(&w[1]), Some(std::cmp::Ordering::Less)))
        {
            return None;
        }
        Some(Self { xs, ys })
    }

    /// Evaluate at `x`. Outside the knot range this CLAMPS to the nearest knot.
    pub fn interpolate(&self, x: f64) -> f64 {
        if x <= self.xs[0] {
            return self.ys[0];
        }
        if x >= self.xs[self.xs.len() - 1] {
            return self.ys[self.ys.len() - 1];
        }
        // Knots are strictly ascending, so partition_point gives the first
        // index with xs[i] > x; the bracketing pair is (i-1, i).
        let i = self.xs.partition_point(|&k| k <= x);
        let (x0, x1) = (self.xs[i - 1], self.xs[i]);
        let (y0, y1) = (self.ys[i - 1], self.ys[i]);
        y0 + (x - x0) / (x1 - x0) * (y1 - y0)
    }

    /// `(min, max)` of the measured abscissa — the span outside which
    /// [`Curve::interpolate`] clamps.
    pub fn range(&self) -> (f64, f64) {
        (self.xs[0], self.xs[self.xs.len() - 1])
    }

    /// True when `x` falls inside the measured range (i.e. the result of
    /// `interpolate(x)` is interpolated rather than clamped).
    pub fn covers(&self, x: f64) -> bool {
        let (lo, hi) = self.range();
        (lo..=hi).contains(&x)
    }

    /// The knot abscissae.
    pub fn xs(&self) -> &[f64] {
        &self.xs
    }

    /// The knot ordinates.
    pub fn ys(&self) -> &[f64] {
        &self.ys
    }

    /// Number of knots.
    pub fn len(&self) -> usize {
        self.xs.len()
    }

    /// Always false — a `Curve` cannot be constructed empty.
    pub fn is_empty(&self) -> bool {
        false
    }

    /// Resample onto a uniform grid of `n` points spanning `lo..=hi`.
    ///
    /// Convenience for downstream flattening (strata resamples every spectrum
    /// onto a shared lambda grid). Values outside the measured range are
    /// clamped — check [`Curve::covers`] first if that matters.
    pub fn resample(&self, lo: f64, hi: f64, n: usize) -> Vec<f64> {
        if n == 0 {
            return Vec::new();
        }
        if n == 1 {
            return vec![self.interpolate(lo)];
        }
        let step = (hi - lo) / (n - 1) as f64;
        (0..n)
            .map(|i| self.interpolate(lo + step * i as f64))
            .collect()
    }

    /// Parse from a TOML table of the form `{ <x_key> = [...], <y_key> = [...] }`.
    pub fn from_toml(table: &toml::Table, x_key: &str, y_key: &str) -> Option<Self> {
        let xs = float_array(table.get(x_key)?)?;
        let ys = float_array(table.get(y_key)?)?;
        Self::new(xs, ys)
    }

    /// Parse a wavelength curve, accepting every value-column spelling used
    /// on disk: `values`, `n` (dispersion), `intensities` (emission spectra).
    pub fn from_wavelength_toml(table: &toml::Table) -> Option<Self> {
        for y_key in ["values", "n", "intensities"] {
            if table.contains_key(y_key) {
                return Self::from_toml(table, "wavelengths_nm", y_key);
            }
        }
        None
    }

    /// Parse a temperature curve (`{ temps_K = [...], values = [...] }`).
    pub fn from_temp_toml(table: &toml::Table) -> Option<Self> {
        Self::from_toml(table, "temps_K", "values")
    }
}

/// Coerce a TOML array of numbers (ints or floats) to `Vec<f64>`.
pub(crate) fn float_array(value: &toml::Value) -> Option<Vec<f64>> {
    value
        .as_array()?
        .iter()
        .map(|v| v.as_float().or_else(|| v.as_integer().map(|i| i as f64)))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn interpolates_between_knots() {
        let c = Curve::new(vec![400.0, 500.0], vec![1.0, 2.0]).unwrap();
        assert!((c.interpolate(450.0) - 1.5).abs() < 1e-12);
    }

    #[test]
    fn returns_exact_knot_values() {
        let c = Curve::new(vec![400.0, 500.0, 600.0], vec![1.0, 2.0, 3.0]).unwrap();
        assert_eq!(c.interpolate(400.0), 1.0);
        assert_eq!(c.interpolate(500.0), 2.0);
        assert_eq!(c.interpolate(600.0), 3.0);
    }

    #[test]
    fn clamps_outside_range() {
        let c = Curve::new(vec![400.0, 500.0], vec![1.8, 1.7]).unwrap();
        assert_eq!(c.interpolate(100.0), 1.8);
        assert_eq!(c.interpolate(900.0), 1.7);
    }

    #[test]
    fn single_knot_is_constant() {
        let c = Curve::new(vec![420.0], vec![1.82]).unwrap();
        assert_eq!(c.interpolate(1.0), 1.82);
        assert_eq!(c.interpolate(420.0), 1.82);
        assert_eq!(c.interpolate(9000.0), 1.82);
    }

    #[test]
    fn rejects_invalid_knots() {
        assert!(Curve::new(vec![], vec![]).is_none());
        assert!(Curve::new(vec![1.0, 2.0], vec![1.0]).is_none());
        assert!(Curve::new(vec![2.0, 1.0], vec![1.0, 2.0]).is_none());
        // Equal adjacent knots make interpolation ambiguous.
        assert!(Curve::new(vec![1.0, 1.0], vec![1.0, 2.0]).is_none());
    }

    #[test]
    fn range_and_covers() {
        let c = Curve::new(vec![400.0, 600.0], vec![1.0, 2.0]).unwrap();
        assert_eq!(c.range(), (400.0, 600.0));
        assert!(c.covers(500.0));
        assert!(!c.covers(300.0));
        assert!(!c.covers(700.0));
    }

    #[test]
    fn resample_hits_endpoints() {
        let c = Curve::new(vec![0.0, 10.0], vec![0.0, 10.0]).unwrap();
        let g = c.resample(0.0, 10.0, 11);
        assert_eq!(g.len(), 11);
        assert_eq!(g[0], 0.0);
        assert_eq!(g[10], 10.0);
        assert!((g[5] - 5.0).abs() < 1e-12);
    }

    #[test]
    fn many_knots_bracket_correctly() {
        let xs: Vec<f64> = (0..100).map(|i| i as f64).collect();
        let ys: Vec<f64> = (0..100).map(|i| (i * 2) as f64).collect();
        let c = Curve::new(xs, ys).unwrap();
        assert!((c.interpolate(50.5) - 101.0).abs() < 1e-12);
        assert!((c.interpolate(0.5) - 1.0).abs() < 1e-12);
        assert!((c.interpolate(98.5) - 197.0).abs() < 1e-12);
    }
}
